service/tasks/driver.py

```diff
@@ -881,6 +881,29 @@ def deploy_ready_test(driverCls, provider, identity, instance_id,
             celery_logger.debug("Instance IP address missing from : %s." % instance_id)
             raise Exception("Instance IP Missing? %s" % instance_id)
 
+        # A race-condition exists where an instance may enter this task with _two_ private IPs
+        # In this case, only one of them will work, so we need to select the correct port ID
+        if len(instance._node.private_ips) == 2:
+            # Determine which port is the active port
+            core_identity = Identity.objects.get(uuid=core_identity_uuid)
+            network_driver = instance_service._to_network_driver(core_identity)
+            port_id = _select_port_id(network_driver, driver, instance)
+            # NOTE: Strategy is to manage only the first floating IP address. Other IP addresses would be managed by user.
+            floating_ip_addr = selected_floating_ip = None
+            floating_ips = network_driver.list_floating_ips()
+            instance_floating_ips = [fip for fip in floating_ips if fip.get("instance_id",'') == instance_id]
+            selected_floating_ip = instance_floating_ips[0] if instance_floating_ips else None
+            if selected_floating_ip and port_id:
+                floating_ip_addr = selected_floating_ip["floating_ip_address"]
+                previous_port_id = selected_floating_ip["port_id"]
+                if previous_port_id != port_id:
+                    celery_logger.info(
+                            "Re-setting existing floating_ip_address %s port: %s -> %s",
+                        floating_ip_addr, previous_port_id, port_id)
+                    network_driver.neutron.update_floatingip(selected_floating_ip['id'], {'floatingip': {'port_id': port_id}})
```

Comment:
Should deploy ready test be thought of as a question:"are we ready", or should it be: "ensure that we're ready"

Reply:
`deploy_ready_test` was traditionally designed to be an `are we ready` test.

The *bugfix* in question was applied here because it gives users the fastest path back to success (Redeploy, reboot, and your instance will be "fixed")

The code was originally placed in `add_floating_ip/initialize_instance_networking`, but when the instance already has a floating IP (albeit incorrectly attached) that task is not executed.

I'm all 👂 if theres a better way, though..